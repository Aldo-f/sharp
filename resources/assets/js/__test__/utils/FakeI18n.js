export default {
    install() {
        window.i18n = {};
    }
}