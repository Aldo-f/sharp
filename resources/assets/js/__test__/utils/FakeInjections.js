import Vue from 'vue';

const injectedComponents = ['$field', '$form'];

export default {
    provide() {
        return injectedComponents.reduce((injections, prop) => {
            injections[prop] = new Vue();
            return injections;
        }, {});
    }
}