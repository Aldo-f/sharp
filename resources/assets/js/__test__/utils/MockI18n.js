import Localization ,* as localizationFn from '../../mixins/Localization';

const mockLangImplementation = localeKey => `{{ ${localeKey} }}`;

export default {
    install() {
        window.i18n = {};
    },

    mockLangFunction() {
        localizationFn.lang = jest.fn(mockLangImplementation);
        Localization.methods.l = jest.fn(mockLangImplementation);
    }
}