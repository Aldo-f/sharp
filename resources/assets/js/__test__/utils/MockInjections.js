import Vue from 'vue';

const injectedComponents = {
    $field:{},
    $form:{
        data: () => ({
            errors: {}
        })
    },
    $tab:{}
};

export default {
    provide() {
        return Object.keys(injectedComponents).reduce((injections, compName) => {
            let Comp = Vue.extend(injectedComponents[compName]);
            injections[compName] = new Comp();
            return injections;
        }, {});
    }
}