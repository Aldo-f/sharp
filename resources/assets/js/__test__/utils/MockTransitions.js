export default {
    install() {
        window.getComputedStyle = jest.fn(() => ({
            animationDelay:'',
            animationDuration:'',
            transitionDelay:'',
            transitionDuration:''
        }));
    }
}