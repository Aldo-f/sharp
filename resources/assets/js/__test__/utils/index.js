export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC } from './mockSFC';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}