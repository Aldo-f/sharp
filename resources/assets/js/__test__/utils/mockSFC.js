export function mockSFC (sfcModule, { template, render, ...extend }={}, mockOptions={}) {

    let { isJsFile } = mockOptions;

    if(isJsFile) {
        sfcModule.__oldDefault__ = { ...sfcModule.default }
    }

    Object.assign(sfcModule.default, {
        mixins: [],
        'extends': extend,
        template: template,
        render: !template ? render || function() { return this._v(`__MOCKED_SFC_${sfcModule.name}__\n`) } : null,
        watch: {},
        beforeCreate: ()=>{},
        created: ()=>{},
        beforeMount: ()=>{},
        mounted: ()=>{},
        beforeUpdate: ()=>{},
        updated: ()=>{},
        beforeDestroy: ()=>{},
        destroyed: ()=>{},
        data: ()=>({}),
        provide: {},
        inject: [],
    });

    //console.log(sfcModule);
    return sfcModule.default;
}

export function unmockSFC(sfcModule) {
    if(sfcModule.__oldDefault__) {
        sfcModule.default = sfcModule.__oldDefault__;
        return;
    }

    let { 'default':defaultObj, ...options } = sfcModule;
    sfcModule.default = options;
}