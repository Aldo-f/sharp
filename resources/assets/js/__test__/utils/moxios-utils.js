import moxios from 'moxios';

export function nextRequestFulfilled(config) {
    return new Promise(resolve => {
        moxios.wait(async () => {
            let request = moxios.requests.mostRecent();
            let response = await request.respondWith(config);
            resolve(response); // moxios Response object
        })
    });
}
