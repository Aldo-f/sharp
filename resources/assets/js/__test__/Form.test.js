import Vue from 'vue';
import Form from '../components/form/Form.vue';

import * as fieldDisplayModule from '../components/form/field-display/FieldDisplay';
import * as gridModule from '../components/Grid.vue';
import * as fieldsLayoutModule from '../components/form/FieldsLayout.vue';
import * as tabbedLayoutModule from '../components/TabbedLayout.vue';

import { ErrorNode } from '../mixins';

import { mockSFC, unmockSFC } from "./utils";

import moxios from 'moxios';
import {MockInjections, MockI18n} from "./utils";
import { nextRequestFulfilled } from './utils/moxios-utils';

describe('sharp-form', ()=>{
    Vue.use(MockI18n);
    MockI18n.mockLangFunction();

    Vue.component('sharp-form', Form);
    Vue.component('sharp-field-display', {
        name: 'SharpFieldContainer',
        inheritAttrs: false,
        mixins: [ ErrorNode ],
        template: '<div> FIELD DISPLAY MOCK </div>'
    });

    beforeAll(()=>{
        mockSFC(gridModule,{
            template: `
            <div id="MOCKED_GRID">
                <slot v-bind="rows[0][0]"></slot>
            </div>
            `
        });

        mockSFC(fieldsLayoutModule, {
            template: `
            <div id="MOCKED_FIELDS_LAYOUT">
                <slot v-bind="layout[0][0]"></slot>
            </div>
            `
        });

        mockSFC(tabbedLayoutModule, {
            template:`
            <div id="MOCKED_TABBED_LAYOUT">
                <slot v-bind="layout.tabs[0]"></slot>
            </div>
            `
        });
    });

    afterAll(()=>{
        unmockSFC(gridModule);
        unmockSFC(tabbedLayoutModule);
        unmockSFC(fieldsLayoutModule);
    });

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-form 
                    entity-key="spaceship" 
                    instance-id="10"
                    :independant="independant" 
                    :props="props"
                    :ignore-authorizations="ignoreAuthorizations">
                </sharp-form>
            </div>
        `;
    });

    afterEach(()=>{
        moxios.uninstall();
    });

    it('can mount sharp-form', async ()=>{
        await createVm();

        await nextRequestFulfilled({
            status: 200,
            response: {
                data:{
                    title: null
                },
                fields: {
                    title: {
                        type: 'text'
                    }
                },

                layout: {
                    tabs : [
                        {
                            columns: [
                                {
                                    fields: [
                                        [{ key: 'title'}]
                                    ]
                                }
                            ]
                        }
                    ]
                }
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can mount "independant" sharp-form', async ()=>{
        await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            type: 'text'
                        }
                    },

                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can mount sharp-form with alert', async ()=>{
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            type: 'text'
                        }
                    },

                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        });

        $form.errors = {
            title: 'error'
        };

        await Vue.nextTick();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [MockInjections, customOptions],

        props: ['independant', 'props', 'ignoreAuthorizations'],

        created() {
            let { axiosInstance } = this._provided;
            moxios.install(axiosInstance);
            moxios.uninstall = moxios.uninstall.bind(moxios, axiosInstance);
        }
    });

    await Vue.nextTick();

    return vm.$children[0];
}