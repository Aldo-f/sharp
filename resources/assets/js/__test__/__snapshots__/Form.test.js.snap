// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sharp-form can mount "independant" sharp-form 1`] = `
<div id="app">
    <div class="SharpForm">
        <div role="alert" class="SharpNotification SharpNotification__inline SharpNotification__inline--error" style="display: none;">
            <div class="SharpNotification__details">
                <div class="SharpNotification__text-wrapper">
                    <p class="SharpNotification__title"></p>
                    <p class="SharpNotification__subtitle"></p>
                </div>
            </div>
        </div>
        <div id="MOCKED_TABBED_LAYOUT">
            <div id="MOCKED_GRID">
                <div id="MOCKED_FIELDS_LAYOUT">
                    <div> FIELD DISPLAY MOCK </div>
                </div>
            </div>
        </div>
    </div>
</div>
`;

exports[`sharp-form can mount sharp-form 1`] = `
<div id="app">
    <div class="SharpForm">
        <div role="alert" class="SharpNotification SharpNotification__inline SharpNotification__inline--error" style="display: none;">
            <div class="SharpNotification__details">
                <div class="SharpNotification__text-wrapper">
                    <p class="SharpNotification__title"></p>
                    <p class="SharpNotification__subtitle"></p>
                </div>
            </div>
        </div>
        <div id="MOCKED_TABBED_LAYOUT">
            <div id="MOCKED_GRID">
                <div id="MOCKED_FIELDS_LAYOUT">
                    <div> FIELD DISPLAY MOCK </div>
                </div>
            </div>
        </div>
    </div>
</div>
`;
